Modules/_threadmodule.c

```diff
@@ -2590,42 +2647,34 @@ _thread_set_name_impl(PyObject *module, PyObject *name_obj)
     PyInterpreterState *interp = _PyInterpreterState_GET();
     const char *encoding = interp->unicode.fs_codec.encoding;
 #endif
-    PyObject *name_encoded;
-    name_encoded = PyUnicode_AsEncodedString(name_obj, encoding, "replace");
-    if (name_encoded == NULL) {
-        return NULL;
-    }
 
-#ifdef _PYTHREAD_NAME_MAXLEN
-    // Truncate to _PYTHREAD_NAME_MAXLEN bytes + the NUL byte if needed
-    if (PyBytes_GET_SIZE(name_encoded) > _PYTHREAD_NAME_MAXLEN) {
-        PyObject *truncated;
-        truncated = PyBytes_FromStringAndSize(PyBytes_AS_STRING(name_encoded),
-                                              _PYTHREAD_NAME_MAXLEN);
-        if (truncated == NULL) {
-            Py_DECREF(name_encoded);
+    int rc = set_encoded_thread_name(name_obj, encoding);
+    if (rc == -1) {
+        /* Confirm a Python exception was set by the helper.
+           If not, convert to a runtime error (defensive). */
+        if (PyErr_Occurred()) {
             return NULL;
         }
-        Py_SETREF(name_encoded, truncated);
+        PyErr_SetString(PyExc_RuntimeError, "internal error in set_encoded_thread_name");
+        return NULL;
```

Comment:
If `errno=-1` was set by `pthread_setname_np()` or other functions (very unlikely), it should be treated as any other `errno` value. I.e. fallback to raising OSError instead of RuntimeError.

So just write
```c
    if (rc == -1 && PyErr_Occurred()) {
        return NULL;
    }
```
Most likely all `errno` values set by library functions are positive, so `PyErr_Occurred()` may be not needed, but it will not harm to check.